doc/decisions/0_drafts/errors_core.md

```diff
@@ -0,0 +1,222 @@
+  For example, only `NULL` is an invalid pointer and sometimes even that is a legitimate non-error result.
+- It should be easy to detect error cases and distinguish between errors, once an error is detected.
+- It should be hard to miss the fact that errors are possible.
+  Note, however, this does not mean checking for an error should be enforced.
+  There are many cases, where the caller knows an error is impossible, because of the previous code path.
+  The requirement is analogous to the `Key * errorKey` argument.
+  The presence of the argument makes it obvious that errors _can_ happen, but you can easily omit the error check when you want to.
+- Adding error indications should not make the API less convenient to use.
+
+## Assumptions
+
+- In many cases, it is possible to avoid error cases through the preceding code path.
+- When an error occurred, the kind error can be distinguished after the fact in many cases.
+- Callers normally don't care about the exact error, unless they can do something about it.
+  This could be:
+  - reporting the error to a user
+  - switching to a different code path
+  - (during debugging) changing the code
+- Functions can be written in a such a way that they detect and report all errors without permanent effects.
+
+## Considered Alternatives
+
+### Extra argument
+
+All functions that report errors take a `int * error` argument to which errors can be written.
+
+This makes the API much more convoluted to use.
+
+A benefit would be the return value remains fully usable.
+
+### Thread Local Storage
+
+Use a field in thread local storage (TLS) to store errors similar to `errno`.
+
+This leaves the API design entirely open.
+
+TLS was only standardized in C11, while Elektra currently only requires C99.
+
+This has all the same issues as `errno`: needs to be cleared to detect errors, can only store one error (per thread), etc.
+
+### Pragmatic solution
+
+The pragmatic solution here is to simply assume "everything will be fine".
+
+Many functions only return errors for simple precondition violations, like passing a `NULL` pointer or otherwise invalid argument.
+Especially for `libelektra-core` (the only place where `Key * errorKey` truly cannot be used), all functions must be side effect free.
+Therefore, the only possible cause of errors are invalid arguments.
+
+If we now assume the function is written such that it fails before it changes any argument's data, then we can infer that only a single error indicating return value is needed.
+Because the arguments are unchanged and the only possible cause of errors, we can distinguish between errors after the fact.
+
+As stated, many functions' only error cases are precondition violations.
+If the preconditions are simple enough, they can be checked beforehand if it is not clear whether the precondition holds.
+
+Take for example some hypothetical `keyValue` function, in a world where we have a `key->value` container for COW that should always be present.
```

Comment:
I'll try to come up with something. But it will probably still be hypothetical. I wanted simple cases, where we can focus more on the preconditions and less on what the function does. But I also didn't want to use something, where we're already planning to change the API.